substrate-node/pallets/pallet-minting/src/lib.rs

```diff
@@ -0,0 +1,468 @@
+            .into_iter()
+            .filter(|period| {
+                match Self::payout_period(&node, node.farm_id, node.connection_price, period) {
+                    Ok(_) => true,
+                    Err(e) => {
+                        log::debug!("payout failed: {:?}", e);
+                        false
+                    }
+                }
+            })
+            .collect();
+        PayablePeriods::<T>::insert(node_id, payable_periods);
+
+        Ok(().into())
+    }
+
+    pub fn payout_period(
+        node: &pallet_tfgrid::TfgridNode<T>,
+        farm_id: u32,
+        connection_price: u32,
+        period: &types::NodePeriodInformation,
+    ) -> DispatchResultWithPostInfo {
+        log::debug!("min capacity: {:?}", period.min_capacity);
+        let (cu, su) = period.min_capacity.cloud_units_permill();
+        log::debug!("cu: {}", cu);
+        log::debug!("su: {}", su);
+        let period_length = T::PeriodTreshold::get();
+
+        let farming_policy = pallet_tfgrid::FarmingPoliciesMap::<T>::get(period.farming_policy);
+        let farm = pallet_tfgrid::Farms::<T>::get(farm_id)
+            .ok_or(pallet_tfgrid::Error::<T>::FarmNotExists)?;
+        let farm_twin = pallet_tfgrid::Twins::<T>::get(farm.twin_id)
+            .ok_or(pallet_tfgrid::Error::<T>::TwinNotExists)?;
+
+        let mut cu_reward = 0;
+        let mut su_reward = 0;
+
+        let uptime_percentage = (period.uptime * 1_000 / period_length) as u128;
+
+        if farming_policy.id == 1 || farming_policy.id == 2 {
+            // CU/SU rewards are the amount of CU/SU for the minimal capacity scaled on a period
+            cu_reward = cu * farming_policy.cu as u64;
+            su_reward = su * farming_policy.su as u64;
+        } else {
+            let minimal_uptime_met =
+                period.uptime < (period_length * farming_policy.minimal_uptime as u64) / 1000;
+            if !minimal_uptime_met {
+                return Err(Error::<T>::UptimeNotMetInPeriodFollowingSla.into());
+            }
+        }
+
+        log::debug!("cu reward: {}", cu_reward);
+        log::debug!("su reward: {}", su_reward);
+
+        // Network traffic rewards are per Gigabyte for a period
+        let nru_reward = (period.nru as u128 * ONE_MILL / GIGABYTE) * farming_policy.nu as u128;
+        log::debug!("nru reward: {}", nru_reward);
+
+        // Public IP rewards are per public ip per hour for a period
+        let ipu_reward = period.ipu * farming_policy.ipv4 as u128 / 3600;
+        log::debug!("ipu reward: {}", ipu_reward);
+
+        let mut base_payout =
+            (cu_reward as u128 + su_reward as u128 + nru_reward) / ONE_MILL + ipu_reward;
+        log::debug!("total musd reward: {}", base_payout);
+        // Caculate actual payout for the period uptime
+        base_payout = base_payout * uptime_percentage / 1_000;
+        // Inflate base payout for more precision
+        base_payout = base_payout * 100_000_000;
```

Comment:
I am not sure why a precision trick is required here
I don t see where it is divided back

Reply:
payouts in musd are converted to units TFT (1e7 decimals precision)